### DISCUSSION TITLE
How to use a callback

### DISCUSSION THREAD

**Discussion by diegoarcega** [TanStack/query#745] [General]
I want to do something like this
https://github.com/slorber/react-async-hook#usecase-injecting-async-feedback-into-buttons

I need a callback returned from the use query hook. When I click on a list of products, I need to request data and then redirect the user to a new page. 

In the docs this is not explained or not explained very well.

**Comment by wolverineks** [comment]
https://react-query.tanstack.com/docs/api#usequery

use `isLoading` or `status === 'loading'` for first load
use `isFetching` for manual refetching

**Reply by diegoarcega** [reply]
this is not what I am looking for, please check the useAsyncCallback hook that I linked above.

Imagine you have a list of products and when the user clicks on a product the app fetches some data from the server and then redirects the user to a new page that doesn't have the id of the product the user clicked.

**Reply by cherniavskii** [reply]
Sounds like you want to prefetch query.
See Prefetching docs section - https://react-query.tanstack.com/docs/guides/prefetching

**Reply by diegoarcega** [reply]
the id of the product is unknown before the user clicks on the product, so I cant prefetch

**Reply by cherniavskii** [reply]
But you can prefetch after user clicks, right?

**Reply by diegoarcega** [reply]
I can, but it only makes sense when I have a useQuery that will 'consume' the prefetched data, and currently I dont have.

**Reply by wolverineks** [reply]
could you do something like
```ts
const product = useQuery({ 
  ..., 
  config: { 
    onSuccess: ({ id }: Product) => redirect('/products/' + id), 
    enabled: false 
})

<button disabled={product.isLoading} onClick={() => product.refetch()}>Go To Product</button>
```

**Reply by diegoarcega** [reply]
hm I will try that, thank you

**Reply by catchshyam** [reply]
@diegoarcega did you manage to find a solution?
